lib/scholar/metrics/distance.ex

```diff
@@ -218,47 +317,73 @@ defmodule Scholar.Metrics.Distance do
   1 - \\frac{u \\cdot v}{\\|u\\|_2 \\|v\\|_2}
   $$
 
+  ## Options
+
+    * `:axes` - Axes to calculate the distance over. By default the distance
+      is calculated between the whole tensors.
+
   ## Examples
 
       iex> x = Nx.tensor([1, 2])
       iex> y = Nx.tensor([5, 2])
       iex> Scholar.Metrics.Distance.cosine(x, y)
       #Nx.Tensor<
         f32
-        0.2525906562805176
-      >
-
-      iex> x = Nx.tensor([1, 2])
-      iex> y = Nx.tensor([1, 2])
-      iex> Scholar.Metrics.Distance.cosine(x, y)
-      #Nx.Tensor<
-        f32
-        0.0
+        0.25259071588516235
       >
 
       iex> x = Nx.tensor([1, 2])
       iex> y = Nx.tensor([1, 2, 3])
       iex> Scholar.Metrics.Distance.cosine(x, y)
       ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+
+      iex> x = Nx.tensor([[1, 2, 3], [0, 0, 0], [5, 2, 4]])
+      iex> y = Nx.tensor([[1, 5, 2], [2, 4, 1], [0, 0, 0]])
+      iex> Scholar.Metrics.Distance.cosine(x, y, axes: [1])
+      #Nx.Tensor<
+        f32[3]
+        [0.1704850196838379, 1.0, 1.0]
+      >
   """
-  @spec cosine(Nx.t(), Nx.t()) :: Nx.t()
-  defn cosine(x, y) do
+  @spec cosine(Nx.t(), Nx.t(), keyword()) :: Nx.t()
+  defn cosine(x, y, opts \\ []) do
+    # Detect very small values that could lead to surprising
+    # results and numerical stability issues. Every value smaller
+    # than `cutoff` is considered small
+    cutoff = 10 * 2.220446049250313e-16
     assert_same_shape!(x, y)
-    norm_x = Nx.LinAlg.norm(x)
-    norm_y = Nx.LinAlg.norm(y)
 
-    cond do
-      norm_x == 0.0 and norm_y == 0.0 ->
-        0.0
+    opts = keyword!(opts, [:axes])
 
-      norm_x == 0.0 or norm_y == 0.0 ->
-        1.0
+    x_squared = x * x
+    y_squared = y * y
 
-      true ->
-        numerator = Nx.dot(x, y)
-        denominator = norm_x * norm_y
-        1.0 - numerator / denominator
-    end
+    norm_x =
+      x_squared
+      |> Nx.sum(axes: opts[:axes], keep_axes: true)
+      |> Nx.sqrt()
+
+    norm_y =
+      y_squared
+      |> Nx.sum(axes: opts[:axes], keep_axes: true)
+      |> Nx.sqrt()
+
+    norm_x_mask = norm_x |> Nx.squeeze()
+
+    norm_y_mask = norm_y |> Nx.squeeze()
+
+    both_zero? = norm_x_mask == 0.0 and norm_y_mask == 0.0
+    one_zero? = Nx.logical_xor(norm_x_mask == 0.0, norm_y_mask == 0.0)
+
+    norm_x = Nx.select(norm_x > cutoff, norm_x, 1.0)
+    normalized_x = x / norm_x
+
+    norm_y = Nx.select(norm_y > cutoff, norm_y, 1.0)
+    normalized_y = y / norm_y
```

Comment:
We can move this before we compute masks, so then we can just do `norm_x = Nx.squeeze(norm_x, axes: opts[:axes])` and use this.